File: II-145-3/VendingMachine.cs
using System;
using System.Collections;
using System.Globalization;
using System.Text;
using System.Text.RegularExpressions;

public class VendingMachine
{
	public int motorUse( string[] prices, string[] purchases )
	{
		int result = 0;
		int time = 0;
		bool first = true;
		int m = prices.Length;
		int n = prices[0].Split().Length;
		int[, ] M = new int[m, n];
		int[] sum = new int[n];
		for( int i = 0; i < m; i++ )
		{
			string[] x = prices[i].Split();
			for( int j = 0; j < n; j++ )
			{
				M[i, j] = int.Parse( x[j] );
				sum[j] += M[i, j];
			}
		}
		int curCol = 0;
		int c = 0; int max = 0;
		for( int step = 0; step < purchases.Length; step++ )
		{
			string[] x = purchases[step].Split( ",:".ToCharArray() );
			int row = int.Parse( x[0] );
			int col = int.Parse( x[1] );
			int t = int.Parse( x[2] );
			if ( first || t - time >= 5 ) 
			{
				c = 0;
				max = sum[0];				
				for( int i = 1; i < n; i++ )
				{
					if ( max < sum[i] )
					{
						max = sum[i]; c = i;
					}
				}
				result += Math.Min( Math.Abs( curCol - c ), n - Math.Abs( curCol - c ) );
				curCol = c;
			}
			first = false;
			if ( M[row, col] == 0 ) return -1;
			result += Math.Min( Math.Abs( curCol - col ), n - Math.Abs( curCol - col ) );
			sum[col] -= M[row, col];
			M[row, col] = 0;
			time = t;
			curCol = col;
		}
		c = 0;
		max = sum[0];				
		for( int i = 1; i < n; i++ )
		{
			if ( max < sum[i] )
			{
				max = sum[i]; c = i;
			}
		}
		result += Math.Min( Math.Abs( curCol - c ), n - Math.Abs( curCol - c ) );
		return result;
	}

	NumberFormatInfo nfi = new CultureInfo( "en-US", false ).NumberFormat;

// BEGIN CUT HERE
	public static void Main(string[] args)
	{
		DateTime t1 = DateTime.Now;
		eq(1,(new VendingMachine()).motorUse(new String[] {"100 100 100"}, new String[] {"0,0:0", "0,2:5", "0,1:10"}),4);
		eq(2,(new VendingMachine()).motorUse(new String[] {"100 200 300 400 500 600"}, new String[] {"0,2:0", "0,3:5", "0,1:10", "0,4:15"}),17);
		eq(3,(new VendingMachine()).motorUse(new String[] {"100 200 300 400 500 600"}, new String[] {"0,2:0", "0,3:4", "0,1:8", "0,4:12"}),11);
		eq(4,(new VendingMachine()).motorUse(new String[] {"100 100 100"}, new String[] {"0,0:10", "0,0:11"}),-1);
		eq(5,(new VendingMachine()).motorUse(new String[] {"100 200 300",
"600 500 400"}, new String[] {"0,0:0", "1,1:10", "1,2:20",
 "0,1:21", "1,0:22", "0,2:35"}),6);
		DateTime t2 = DateTime.Now;
		TimeSpan t = t2 - t1;
		Console.WriteLine( "{0} sec", t.TotalMilliseconds / 1000 ); 
	}
	private static void eq(int n, int a, int b)
	{
		if(a == b)
			Console.WriteLine("Case "+n+" passed.");
		else
			Console.WriteLine("Case "+n+" failed: expected \""+b+"\", received \""+a+"\".");
	}
	private static void eq(int n, double a, double b)
	{
		if(a == b)
			Console.WriteLine("Case "+n+" passed.");
		else
			Console.WriteLine("Case "+n+" failed: expected \""+b+"\", received \""+a+"\".");
	}
	private static void eq(int n, char a, char b)
	{
		if(a == b)
			Console.WriteLine("Case "+n+" passed.");
		else
			Console.WriteLine("Case "+n+" failed: expected '"+b+"', received '"+a+"'.");
	}
	private static void eq(int n, long a, long b)
	{
		if(a == b)
			Console.WriteLine("Case "+n+" passed.");
		else
			Console.WriteLine("Case "+n+" failed: expected \""+b+"\", received \""+a+"\".");
	}
	private static void eq(int n, bool a, bool b)
	{
		if(a == b)
			Console.WriteLine("Case "+n+" passed.");
		else
			Console.WriteLine("Case "+n+" failed: expected \""+b+"\", received \""+a+"\".");
	}
	private static void eq(int n, string a, string b)
	{
		if(a == b)
			Console.WriteLine("Case "+n+" passed.");
		else
			Console.WriteLine("Case "+n+" failed: expected \""+b+"\", received \""+a+"\".");
	}
	private static void eq(int n, int[] a, int[] b)
	{
		if(a.Length != b.Length)
		{
			Console.WriteLine("Case "+n+" failed: returned "+a.Length+" elements; expected "+b.Length+" elements.");
			return;
		}
		for(int i= 0; i < a.Length; i++)
			if(a[i] != b[i])
			{
				Console.WriteLine("Case "+n+" failed. Expected and returned array differ in position "+i);
				print(b);
				print(a);
				return;
			}
		Console.WriteLine("Case "+n+" passed.");
	}
	private static void eq(int n, double[] a, double[] b)
	{
		if(a.Length != b.Length)
		{
			Console.WriteLine("Case "+n+" failed: returned "+a.Length+" elements; expected "+b.Length+" elements.");
			return;
		}
		for(int i= 0; i < a.Length; i++)
			if(a[i] != b[i])
			{
				Console.WriteLine("Case "+n+" failed. Expected and returned array differ in position "+i);
				print(b);
				print(a);
				return;
			}
		Console.WriteLine("Case "+n+" passed.");
	}
	private static void eq(int n, long[] a, long[] b)
	{
		if(a.Length != b.Length)
		{
			Console.WriteLine("Case "+n+" failed: returned "+a.Length+" elements; expected "+b.Length+" elements.");
			return;
		}
		for(int i= 0; i < a.Length; i++)
			if(a[i] != b[i])
			{
				Console.WriteLine("Case "+n+" failed. Expected and returned array differ in position "+i);
				print(b);
				print(a);
				return;
			}
		Console.WriteLine("Case "+n+" passed.");
	}
	private static void eq(int n, string[] a, string[] b)
	{
		if(a.Length != b.Length)
		{
			Console.WriteLine("Case "+n+" failed: returned "+a.Length+" elements; expected "+b.Length+" elements.");
			return;
		}
		for(int i= 0; i < a.Length; i++)
			if(a[i] != b[i])
			{
				Console.WriteLine("Case "+n+" failed. Expected and returned array differ in position "+i);
				print(b);
				print(a);
				return;
			}
		Console.WriteLine("Case "+n+" passed.");
	}
	private static void print(int a)
	{
		Console.Write(a+" ");
	}
	private static void print(string s)
	{
		Console.Write("\""+s+"\" ");
	}
	private static void print(int[] rs)
	{
		if(rs == null) return;
		Console.Write('{');
		for(int i= 0; i < rs.Length; i++)
		{
			Console.Write(rs[i]);
			if(i != rs.Length-1)
				Console.Write(", ");
		}
		Console.WriteLine('}');
	}
	private static void print(double[] rs)
	{
		if(rs == null) return;
		Console.Write('{');
		for(int i= 0; i < rs.Length; i++)
		{
			Console.Write(rs[i]);
			if(i != rs.Length-1)
				Console.Write(", ");
		}
		Console.WriteLine('}');
	}
	private static void print(long[] rs)
	{
		if(rs == null) return;
		Console.Write('{');
		for(int i= 0; i < rs.Length; i++)
		{
			Console.Write(rs[i]);
			if(i != rs.Length-1)
				Console.Write(", ");
		}
		Console.WriteLine('}');
	}
	private static void print(string[] rs)
	{
		if(rs == null) return;
		Console.Write('{');
		for(int i= 0; i < rs.Length; i++)
		{
			Console.Write("\""+rs[i]+"\"");
			if(i != rs.Length-1)
				Console.Write(", ");
		}
		Console.WriteLine('}');
	}
	private static void nl()
	{
		Console.WriteLine();
	}

// END CUT HERE
}

File: II-145-3/VendingMachine.cpp
#include <algorithm>
#include <cmath>
#include <cstdlib>
#include <functional>
#include <iostream>
#include <iterator>
#include <map>
#include <numeric>
#include <set>
#include <sstream>
#include <stack>
#include <string>
#include <queue>
#include <vector>
using namespace std;
// BEGIN CUT HERE
#include <Windows.h>
#pragma comment( lib, "Winmm" )
// END CUT HERE
typedef vector<int> vi;
typedef vector<vi> vvi;
typedef vector<string> vs;
#define fora(i,v) for(int i = 0; i < (v).size(); i++)
#define fora2(i,j,v) fora(i,v) fora(j,v[i])
#define all(v) (v).begin(), (v).end()
template<typename T, typename Predicate> vector<T> Parse( string s, Predicate predicate )
{
	vector<T> result; bool isWord = false; int j = 0;
	for( int i = 0; i < s.size(); i++ )
		if ( isWord )
		{
			if ( !predicate( s[i] ) )
			{
				istringstream S( s.substr( j, i - j ) );
				T current; S >> current; 
				if ( S.rdstate() ^ ios::failbit )
					result.push_back( current );
				isWord = false;
			}
		}
		else
		{
			if ( predicate( s[i] ) )
			{
				j = i; isWord = true;
			}
		}
		if ( isWord )
		{
			istringstream S( s.substr( j ) );
			T current; S >> current; 
			if ( S.rdstate() ^ ios::failbit )
				result.push_back( current );
		}
		return result;
}
template<typename Predicate> vs parse( string s, Predicate predicate, bool skipEmpty = true )
{
	vs result; bool isWord = false; int j = 0;
	for( int i = 0; i < s.size(); i++ )
		if ( isWord )
		{
			if ( !predicate( s[i] ) )
			{
				result.push_back( s.substr( j, i - j ) ); isWord = false;
			}
		}
		else
		{
			if ( predicate( s[i] ) )
			{
				j = i; isWord = true;
			} else if ( !skipEmpty )
				result.push_back( "" );
		}
		if ( isWord )
			result.push_back( s.substr( j ) );
		else if ( !skipEmpty && !s.empty() )
			result.push_back( "" );
		return result;
}
struct split_helper: unary_function<int, bool>
{
	explicit split_helper( string delimeter ): delimeter(delimeter) {};
	bool operator() ( int c ) const
	{
		return delimeter.find( c ) == string::npos;
	}
private: string delimeter;
};
template <typename T> vector<T> Split( string s, string delimeter = " " )
{
	return Parse<T>( s, split_helper( delimeter ) );
}
vs split( const string& s, const string& delimeter = " ", bool skipEmpty = true )
{
	return parse( s, split_helper( delimeter ), skipEmpty );
}
struct VendingMachine
{
	int motorUse( vs prices, vs purchases )
	{
		int result = 0; set<string> S;
		fora(i, purchases)
			if ( !S.insert( split( purchases[i], ":" )[0] ).second )
				return -1;
		vvi P;
		fora(i, prices)
			P.push_back( Split<int>( prices[i] ) );
		int n = P[0].size(); vi A(n, 0);
		fora2(i, j, P)
			A[j] += P[i][j];
		int time = -5, curCol = 0;
		fora(i, purchases)
		{
			vi z = Split<int>( purchases[i], ",:" );
			if ( z[2] - time >= 5 )
			{
				int m = distance( A.begin(), max_element( all(A) ) );
				result += min( abs(curCol - m), n - abs(curCol - m) );
				curCol = m;
			}
			A[z[1]] -= P[z[0]][z[1]];
			result += min( abs(curCol - z[1]), n - abs(curCol - z[1]) );
			curCol = z[1];
			time = z[2];
		}
		int m = distance( A.begin(), max_element( all(A) ) );
		result += min( abs(curCol - m), n - abs(curCol - m) );
		return result;
	}
	// BEGIN CUT HERE
	private:
	template <typename T> string print_array( const vector<T>& v ) { ostringstream os; os << "{ "; for( typename vector<T>::const_iterator it = v.begin(); it != v.end(); ++it ) os << '\"' << *it << "\","; os << " }"; return os.str(); }
	template<typename T> void verify_case( int Case, const vector<T>& expected, const vector<T>& received ) { cerr << "Test Case #" << Case << "..."; if ( expected == received ) cerr << "PASSED. "; else { cerr << "FAILED." << endl; cerr << "\tExpected: " << print_array( expected ) << endl; cerr << "\tReceived: " << print_array( received ) << endl; } }
	void verify_case( int Case, const long long& expected, const long long& received ) { cerr << "Test Case #" << Case << "..."; if ( expected == received ) cerr << "PASSED. "; else { cerr << "FAILED." << endl; cerr << "\tExpected: \"" << expected << '\"' << endl; cerr << "\tReceived: \"" << received << '\"' << endl; } }
	void verify_case( int Case, const double& expected, const double& received ) { cerr << "Test Case #" << Case << "..."; if ( expected == received ) cerr << "PASSED. "; else { cerr << "FAILED." << endl; cerr << "\tExpected: \"" << expected << '\"' << endl; cerr << "\tReceived: \"" << received << '\"' << endl; } }
	void verify_case( int Case, const int& expected, const int& received ) { cerr << "Test Case #" << Case << "..."; if ( expected == received ) cerr << "PASSED. "; else { cerr << "FAILED." << endl; cerr << "\tExpected: \"" << expected << '\"' << endl; cerr << "\tReceived: \"" << received << '\"' << endl; } }
	void verify_case( int Case, const string& expected, const string& received ) { cerr << "Test Case #" << Case << "..."; if ( expected == received ) cerr << "PASSED. "; else { cerr << "FAILED." << endl; cerr << "\tExpected: \"" << expected << '\"' << endl; cerr << "\tReceived: \"" << received << '\"' << endl; } }
	public:
	void run_test( int Case = -1 ) { vector<vs> A0; vector<vs> A1; vector<int> A2;
	{ /*Test 0*/string _prices[] = 
	{"100 100 100"}
	; vs prices(_prices, _prices + (sizeof(_prices) / sizeof(_prices[0]))); A0.push_back( prices ); string _purchases[] = 
	{"0,0:0", "0,2:5", "0,1:10"}
	; vs purchases(_purchases, _purchases + (sizeof(_purchases) / sizeof(_purchases[0]))); A1.push_back( purchases ); int result = 
	4
	; A2.push_back( result );  }
	{ /*Test 1*/string _prices[] = 
	{"100 200 300 400 500 600"}
	; vs prices(_prices, _prices + (sizeof(_prices) / sizeof(_prices[0]))); A0.push_back( prices ); string _purchases[] = 
	{"0,2:0", "0,3:5", "0,1:10", "0,4:15"}
	; vs purchases(_purchases, _purchases + (sizeof(_purchases) / sizeof(_purchases[0]))); A1.push_back( purchases ); int result = 
	17
	; A2.push_back( result );  }
	{ /*Test 2*/string _prices[] = 
	{"100 200 300 400 500 600"}
	; vs prices(_prices, _prices + (sizeof(_prices) / sizeof(_prices[0]))); A0.push_back( prices ); string _purchases[] = 
	{"0,2:0", "0,3:4", "0,1:8", "0,4:12"}
	; vs purchases(_purchases, _purchases + (sizeof(_purchases) / sizeof(_purchases[0]))); A1.push_back( purchases ); int result = 
	11
	; A2.push_back( result );  }
	{ /*Test 3*/string _prices[] = 
	{"100 100 100"}
	; vs prices(_prices, _prices + (sizeof(_prices) / sizeof(_prices[0]))); A0.push_back( prices ); string _purchases[] = 
	{"0,0:10", "0,0:11"}
	; vs purchases(_purchases, _purchases + (sizeof(_purchases) / sizeof(_purchases[0]))); A1.push_back( purchases ); int result = 
	-1
	; A2.push_back( result );  }
	{ /*Test 4*/string _prices[] = 
	{"100 200 300",
 "600 500 400"}
	; vs prices(_prices, _prices + (sizeof(_prices) / sizeof(_prices[0]))); A0.push_back( prices ); string _purchases[] = 
	{"0,0:0", "1,1:10", "1,2:20",
 "0,1:21", "1,0:22", "0,2:35"}
	; vs purchases(_purchases, _purchases + (sizeof(_purchases) / sizeof(_purchases[0]))); A1.push_back( purchases ); int result = 
	6
	; A2.push_back( result );  }
	for( int i = 0; i < A0.size(); i++ ) if ( Case == -1 || i == Case ) { DWORD start, finish; start = timeGetTime(); verify_case( i, A2[i], motorUse( A0[i], A1[i]) ); finish = timeGetTime(); cout << "Elapsed " << finish - start << " msec" << endl; } }
// END CUT HERE
};
// BEGIN CUT HERE
int main()
{
	DWORD start, finish;
	start = timeGetTime();
	VendingMachine ___test; 
	___test.run_test(); 
	finish = timeGetTime();
	cout << endl << "Elapsed " << finish - start << " msec" << endl;
} 
// END CUT HERE
